src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+                        result = Sse.Add(result, Sse.Multiply(srcVector, srcVector));
+
+                        pSrcCurrent += 4;
+                    }
+
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
+                        result = Sse3.HorizontalAdd(result, result);
+                        result = Sse3.HorizontalAdd(result, result);
+                    }
+                    else
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
+                    }
+
+                    while (pSrcCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                        result = Sse.AddScalar(result, Sse.MultiplyScalar(srcVector, srcVector));
+
+                        pSrcCurrent++;
+                    }
+                }
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+        internal static float SumAbsU(float[] src, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+            Vector128<float> mask;
+
+            if (Sse2.IsSupported)
+            {
+                mask = Sse.StaticCast<int, float>(Sse2.SetAllVector128(0x7FFFFFFF));
+            }
+            else
+            {
+                mask = Sse.SetAllVector128(0x7FFFFFFF);
+            }
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pEnd = psrc + count;
+
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        result = Sse.Add(result, Sse.And(srcVector, mask));
+
+                        pSrcCurrent += 4;
+                    }
+
+                    // TODO: Check code length per line
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
+                        result = Sse3.HorizontalAdd(result, result);
+                        result = Sse3.HorizontalAdd(result, result);
+                    }
+                    else
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
+                    }
+
+                    while (pSrcCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
```

Comment:
I know that this is ported directly from the C++ code. And while I can't exactly state what kind of issues it will cause, it seems wrong/unsafe to be reading memory outside of the bounds the method is given. We don't have any idea what is after the array in memory, but we are reading/loading at least 12 bytes of memory after the array into this Vector128.

Would it make more sense to turn `result` into a float before this line, and then do scalar abs/add at the end of the array?

Reply:
This could cause AV, or any number of other issues.

We should either drop back to scalar or do a double-compute for the last chunk. See https://github.com/dotnet/machinelearning/pull/562/files/f0f81a5019a3c8cbd795a970e40d633e9e1770c1#r204073474 for more details.